### PR TITLE
[Notifier] fix flaky notifier test

### PR DESCRIPTION
**Pull request by smnzhu**
This PR fixes race conditions in `notifier_test`. This test is flaky and often fails, blocking merge.

The race condition in `TestNotifier_AllWorkProcessed` is as follows:

Because we have more producers than consumers, during the first round of notifications the producers will produce 10 notifications, 5 of which will be received immediately by the consumers (since they are already [blocking on read](https://github.com/onflow/flow-go/blob/master/engine/notifier_test.go#L162-L163) from the notification channel). Once these consumers get the notification however, they proceed to [consume from the work queue](https://github.com/onflow/flow-go/blob/master/engine/notifier_test.go#L151-L155), while the remaining 5 notifications will most likely [get dropped](https://github.com/onflow/flow-go/blob/smnzhu/fix-notifier-test/engine/notifier.go#L37-L40) before the consumers have time to come back around to the next iteration of the loop. Therefore, the consumers will never be notified of these items. As the rounds proceed, more and more notifications get dropped. It looks something like below (shrank the example to 2 consumers and 4 producers):

```
Consumer 1 waiting for notification...
Consumer 2 waiting for notification...
Producer 1 pushes to queue and sends notification
Consumer 1 receives notification
Consumer 1 goes to read from queue
Producer 2 pushes to queue and sends notification
Consumer 2 receives notification
Consumer 2 goes to read from work queue
Producer 3 pushes to queue and sends notification (dropped)
Producer 4 pushes to queue and sends notification (dropped)
Producer 1 pushes to queue and sends notification (dropped)
Consumer 1 finishes reading from queue and incrementing counter
Consumer 1 waiting for notification...
Producer 2 pushes to queue and sends notification
Consumer 2 finishes reading from queue and incrementing counter
Consumer 1 receives notification
Consumer 1 goes to read from queue
Consumer 2 waiting for notification...
... 
```

The fix is to just add a small delay in the producers' loop to ensure that consumers are actually listening for notifications by the time the next one comes

This PR also fixes a minor issue in the test:

The error messages are incorrect. The call to [`consumedWork.Load()`](https://github.com/onflow/flow-go/blob/master/engine/notifier_test.go#L169) is evaluated at the time of the function call (at which point it is still 0 since no work has been consumed yet), so the error message will always show up as follows:

```
only consumed 0 units of work but expecting 100
```

What we really need here is a function that returns a boolean indicating whether or not the condition is satisfied eventually, and if not, only then do we create the error message.

### CODE REVIEW

**Review by AlexHentschel — CHANGES_REQUESTED** [review]
I don't think we are on the same page on the specification of this algorithm. The Notifier is supposed to implement this state machine: 
Screen Shot 2021-07-27 at 3 26 52 PM

**Comment by smnzhu** [comment]
> I don't think we are on the same page on the specification of this algorithm. The Notifier is supposed to implement this state machine:

👍 I see that now, however the race condition still remains.

**Comment by AlexHentschel** [comment]
After giving it more thought, I think we have a bit more of a profound problem here. The test actually represents our intended usage pattern. I have illustrated it below: 
Screen Shot 2021-07-27 at 4 01 37 PM

So the test is exactly testing that this implementation is responsive enough. But it seems like it is not (?) So the test failing told us that it took _more_ than 3s (!) to push 100 elements through the `pendingWorkQueue`. The work packages themselfs are empty. So even a _single_ worker thread should be managing this singlehandedly 🤔  ....
But it doesn't seem to. So maybe there is an edge-case in the implementation ... I'll have to think about it a bit more

**Comment by smnzhu** [comment]
@AlexHentschel Thanks for the wonderful diagram 👯 

Two points here:
* I think the test, as you have written it, does not actually do exactly what you think it does. I assume that [this](https://github.com/onflow/flow-go/blob/master/engine/notifier_test.go#L155) break statement is meant to break out of the `for` loop, but actually it only breaks out of the select statement. In VSCode, I actually get an "ineffective break statement. did you mean to break out of the outer loop" compiler warning. I suspect c9956aa might be what you really intended.

**Comment by AlexHentschel** [comment]
Very good point @smnzhu 💡 
* the `break` just stepping out of the `switch` means it keeps repeating to (_busy_) poll

Does that imply that we have problems to get 100 elements through the `pendingWorkQueue` ?!? That seems unlikely. 🤔

**Comment by smnzhu** [comment]
> Does that imply that we have problems to get 100 elements through the `pendingWorkQueue` ?!? That seems unlikely. 🤔

I absolutely agree. At the moment I can't see what could be the problem here :/

**Comment by AlexHentschel** [comment]
I mean, this test is for the _responsive system_. Not sure how overloaded the CI instance gets with all the `Parallel()` tests.

**Comment by smnzhu** [comment]
> I mean, this test is for the _responsive system_. Not sure how overloaded the CI instance gets with all the `Parallel()` tests.

I can try to increase the timeout to 5s, or just remove the `t.Parallel()`, and see if either of those helps?

**Comment by AlexHentschel** [comment]
removing the `t.Parallel()` might help

**Comment by smnzhu** [comment]
> removing the `t.Parallel()` might help

c3bcf2a

**Review by AlexHentschel — APPROVED** [review]
👍 lets give this a try 😅

**Comment by AlexHentschel** [comment]
It is really hard to determine the precise conditions for the notifier test failing. My hypothesis is 
* the test is overloaded as we run this test 100 times in parallel; in a very short time a very large number of go-routines is spawned 
* I would guess that there is some latency spawning new routines (more precisely, I would expect that there is some latency until the go routines actually start their business logic)
  * In the `TestNotifier_AllWorkProcessed` version on master, we just [kick off the go-routines](https://github.com/onflow/flow-go/blob/2d496c1ba83ecd2656fe6861936648b069359023/engine/notifier_test.go#L134-L163) but don't wait for them to spawn and start measuring the time right away. If it takes a bit until go routines actually start their business logic because we are spawning thousands of them in parallel, the 3s wait time might already be over when the routines have started up. The test fails
  * In comparison, the other test `TestNotifier_ManyConsumers` does not seem to fail. Here, we proceed differently: _we wait until the routines have all started up and only then start testing that the system is responsive and processes the load in the expected time

While this is a hypothesis, I managed to reproduce the test failure locally: 
* I increased the number of parallel test executions to https://github.com/onflow/flow-go/blob/2d496c1ba83ecd2656fe6861936648b069359023/engine/notifier_test.go#L173-L175 and added equivalent logic to also spawn 500 test runs for the `TestNotifier_ManyConsumers` in parallel. Then `TestNotifier_AllWorkProcessed` indeed failed. 
* **In [PR #1045](https://github.com/onflow/flow-go/pull/1045)** (targeting this branch), I changed the test slightly: it now waits for all involved go-routines to complete their start up: https://github.com/onflow/flow-go/blob/f89e60aca21a1ff7185d4725266cb9bb4ea4f12a/engine/notifier_test.go#L169 and then starts testing responsiveness. This test succeeds even when I load my machine with 500 test executions of `TestNotifier_AllWorkProcessed` plus 500 `TestNotifier_ManyConsumers` in parallel.

**Comment by AlexHentschel** [comment]
@smnzhu thanks for digging into the test failure. Your investigation and the discussion here were instrumental for me to form my hypothesis on the root cause for the failing test. 

#### Suggestion: merge https://github.com/onflow/flow-go/pull/1045/files and see if it fixes the issue.

File: engine/notifier_test.go [hunk 1]
```diff
@@ -147,30 +144,49 @@ func TestNotifier_AllWorkProcessed(t *testing.T) {
...
+		}
+	}()
+
+	select {
+	case <-time.After(waitFor):
+		return false
+	case <-done:
+		return true
+	}
+}
```
Comment (AlexHentschel) [inline engine/notifier_test.go:192]: Could you explain the reasoning behind adding this method. To me, it looks like it is replicating _exactly_ the functionality of `require.Eventuallyf` (?)

Comment (smnzhu) [reply engine/notifier_test.go:192]: > The call to consumedWork.Load() is evaluated at the time of the function call (at which point it is still 0 since no work has been consumed yet)

In the [old version](https://github.com/onflow/flow-go/blob/master/engine/notifier_test.go#L166-L170), the call to `consumedWork.Load()` is evaluated when `require.Eventuallyf` is called. This occurs **before** most of the actual logic in the goroutines occurs, at which point the value is 0. The call to `require.Eventuallyf` only *returns* after 3 second timeout, but the `consumedWork.Load()` has already been evaluated (and in fact the string interpolation has already happened). 

In other words, the error message "only consumed 0 units of work but expecting 100" has already been created before we even enter the `require.Eventuallyf` function, since it needs to be passed as an argument to the function.

Instead, what we really want is to evaluate `consumedWork.Load()` only **after** the timeout has occured. The only way to do this is to create a function which will tell us whether we timed out waiting for the condition, and then create the error message based on the result of that function.

Comment (smnzhu) [reply engine/notifier_test.go:192]: Perhaps this makes it more clear:

The old code:

```
require.Eventuallyf(t,
	func() bool { return consumedWork.Load() == totalWork },
	3*time.Second, 100*time.Millisecond,
	"only consumed %d units of work but expecting %d", consumedWork.Load(), totalWork,
)
```

is actually equivalent to

```
errorMsg := fmt.Sprintf("only consumed %d units of work but expecting %d", consumedWork.Load(), totalWork)

require.Eventuallyf(t,
	func() bool { return consumedWork.Load() == totalWork },
	3*time.Second, 100*time.Millisecond,
	errorMsg
)
```

Clearly, the error message will not contain the correct value here :)

Comment (zhangchiqing) [reply engine/notifier_test.go:192]: Note, the created notifier has [1 buffer](
https://github.com/onflow/flow-go/blob/58bdf615876666b990388ef3017c1e7badc84745/engine/notifier.go#L31)

so that if all consumers are busy, we can still buffer 1 notification from the producer, so that when the consumer becomes available, it will process the buffered notification. 

Therefore, in your example:
```
Producer 3 pushes to queue and sends notification (dropped)
...
Consumer 1 finishes reading from queue and incrementing counter
Consumer 1 waiting for notification...
```

is actually:
```
Producer 3 pushes to queue and sends notification (buffered)
...
Consumer 1 finishes reading from queue and incrementing counter
Consumer 1 processes the buffered notification
```

Comment (zhangchiqing) [reply engine/notifier_test.go:192]: I'm not sure if that is how `require.Eventually` works. If we look at this source code here, it's similar to what you've implemented.

```golang
func Eventually(t TestingT, condition func() bool, waitFor time.Duration, tick time.Duration, msgAndArgs ...interface{}) bool {
	if h, ok := t.(tHelper); ok {
		h.Helper()
	}

	ch := make(chan bool, 1)

	timer := time.NewTimer(waitFor)
	defer timer.Stop()

	ticker := time.NewTicker(tick)
	defer ticker.Stop()

	for tick := ticker.C; ; {
		select {
		case <-timer.C:
			return Fail(t, "Condition never satisfied", msgAndArgs...)
		case <-tick:
			tick = nil
			go func() { ch <- condition() }()
		case v := <-ch:
			if v {
				return true
			}
			tick = ticker.C
		}
	}
}
```

**Comment by zhangchiqing** [comment]
After reviewing the tests, I think this [fix](https://github.com/onflow/flow-go/pull/1049) might work better.

File: engine/notifier_test.go (hunk 1 above)
Comment (smnzhu) [reply engine/notifier_test.go:192]: But think about what the argument `msgAndArgs` holds when the function is called. Does it hold the value of `consumedWork.Load()` from the future? This is impossible. It holds the current value of `consumedWork.Load()` (which has already executed), which is 0

Comment (smnzhu) [reply engine/notifier_test.go:192]: As for your first point, yes my understanding of the code was not entirely correct, and the original PR summary is no longer accurate.

Comment (zhangchiqing) [reply engine/notifier_test.go:192]: >But think about what the argument msgAndArgs holds when the function is called. Does it hold the value of consumedWork.Load() from the future? This is impossible. It holds the current value of consumedWork.Load() (which has already executed), which is 0

Yes, You are right. I fixed it in my PR https://github.com/onflow/flow-go/pull/1049/commits/974dfb4dfb9cbe90bbd6a22dda1c7f8ff9bea22d

**Comment by AlexHentschel** [comment]
@smnzhu @durkmurder @zhangchiqing 

There has been a lot of thinking around whether or not there are any atomicity guarantees around sending and receiving. To illustrate, lets take a look at the [`TestNotifier_ManyConsumers`](https://github.com/onflow/flow-go/blob/f89e60aca21a1ff7185d4725266cb9bb4ea4f12a/engine/notifier_test.go#L82-L118) test: https://github.com/onflow/flow-go/blob/f89e60aca21a1ff7185d4725266cb9bb4ea4f12a/engine/notifier_test.go#L94-L107
* we have a bunch of workers waiting to receive a notification
* then we have a bunch of producers which send notifications

Could it happen that all the producers override each others notifications and then only a single worker receives one?  Essentially, we are wondering about atomicity guarantees of cannel writes and receives:
* Assuming that there is a consumer waiting to receive a notification and one thread sends a notification into the channel. Is receiving the notification an atomic operations such that any _concurrent_ sending of notifications always observes the latest channel state? 

I think the sad news is that it is not (at least not guaranteed on _all_ hardware platforms) 😭. Lets first look at what is actually guaranteed:
* the go memory model states: 
  > The kth receive on a channel with capacity C happens before the k+Cth send from that channel completes.  
* Note the wording (we will inspect runtime implementation later to provide supporting evidence): 



It is a subtle edge case, which is unfortunately very relevant for our

**Comment by AlexHentschel** [comment]
## Suggestion for next steps. 

1. Lets get the test patched up so CI on `master` stops tripping over the flakey test
2. It would be great to fix the implementation. Unfortunately, as **channels do not provide the required atomicity guarantees we need** (see [argument above](https://github.com/onflow/flow-go/pull/1037#issuecomment-888794269) for details), we have to resort to other concurrency paradigms. A potential solution would be the [`trapdoor` approach](https://github.com/onflow/flow-go/blob/cf75e62397bd7ad5756fdbe45b81851cdef12e5f/engine/consensus/matching/trapdoor.go) (see [Flip 396](https://github.com/onflow/flow/issues/396)). This implementation provides the desired atomicity guarantees.

**Comment by bluesign** [comment]
@AlexHentschel 

>The kth receive on a channel with capacity C happens before the k+Cth send from that channel completes. 

> we don't send so there are no guarantees.

I think you have some wrong conclusion there, it just means channels don't have some `extra buffer` other than capacity. So if you have a channel with capacity 1 you cannot make send call until it is received. This is more like (for C=1) Nth receive happens before N+1 th send 


> We are reading c.qcount outside of the lock (!) Hence, there are no guarantee to observe the most-recent buffer size. If this particular go-routine has previously observed the channel to be full, it might not get the updated value for c.qcount and therefore drop its channel write. There is a mutex further down, but this is only acquired if the routine thinks the channel still has capacity.

Can you explain here a bit more, I checked the go code, I don't see a problem there? Do you mean something like a race condition? 

> channels do not provide the required atomicity guarantees we need 

I don't think there is a problem here, only place to miss notification: 

- before getting in from the channel (before processing notification, so no problem)
 
because on receive there is : 

```go
if !block && (c.dataqsiz == 0 && c.sendq.first == nil \|\|
--
  | c.dataqsiz > 0 && atomicloaduint(&c.qcount) == 0) &&
  | atomicload(&c.closed) == 0 {
  | return
  | }
```

**Comment by smnzhu** [comment]
> Lets get the test patched up so CI on master stops tripping over the flakey test

@AlexHentschel Are we on the same page now regarding https://github.com/onflow/flow-go/pull/1037#discussion_r677981484 ?

I think I will go ahead and take the approach from https://github.com/onflow/flow-go/pull/1049 for `TestNotifier_AllWorkProcessed`, since it appears to solve our problem in a slightly simpler way.

I'll take your changes from https://github.com/onflow/flow-go/pull/1045 for `TestNotifier_ManyConsumers` and the new Readme.

**Comment by durkmurder** [comment]
@AlexHentschel Thank you for this detailed investigation, very well explained. 
I have some thoughts about visible side effects of `qcount` in particular this:

> * We are reading `c.qcount` outside of the lock (!) Hence, there are **no guarantee to observe the most-recent buffer size**. If this particular go-routine has previously observed the channel to be full, it might not get the updated value for `c.qcount` and therefore drop its channel write. There is a mutex further down, but this is only acquired if the routine thinks the channel still has capacity.

I believe that reading outside of lock in this situation behaves like _relaxed-atomic_ because the write is synchronized by mutex. 
There are some hints in source code that imply it:

```
We rely on the side effects of lock release in
	// chanrecv() and closechan() to update this thread's view of c.closed and full().
```

Also in [happens before section of memory model](https://golang.org/ref/mem#tmp_2) there are some hints to that as well.

The only problem I can see is that we get a context switch after entering the `if` statement but before `return false` that will perform a `channrecv` and produce a slot but we are still going to return false. But I am not sure if that's an issue.

**Comment by smnzhu** [comment]
> I believe that reading outside of lock in this situation behaves like *relaxed-atomic* because the write is synchronized by mutex.

This is my understanding as well :)

**Comment by AlexHentschel** [comment]
> I believe that reading outside of lock in this situation behaves like relaxed-atomic because the write is synchronized by mutex.

I agree it is intuitive and out test seems to indicate that for the CI hardware platform our assumption actually holds that `c.qcount` observes the most recent value.

Nevertheless, I still have not managed to construct a proof that `c.qcount` is guaranteed to observes the most recent value. The Go memory model states:

> That is, r is _guaranteed_ to observe w if both of the following hold:
> 1. w happens before r.
> 2. Any other write to the shared variable v either happens before w or after r.
>
> [...] It requires that there are no other writes happening concurrently with w or r.

My understanding is that there might be concurrent writes to `c.qcount`, because we don't hold a mutex while reading it. So my interpretation of the memory model is that there would not be a universal guarantee. 


In the end, this is hard science. The scientific principle says: "If you can't prove it (present hard evidence), you cannot assume it exists".
* A passing test is not satisfactory to guarantee that our approach is correct. We only test with a specific go version on a specific hardware platform. On  different hardware or different compiler versions, the test might fail.

In the end, we _need a proof_ based on the guarantees and specifications of the memory model.

**Comment by AlexHentschel** [comment]
> @AlexHentschel Are we on the same page now regarding #1037 (comment) ?

thanks for the explanation @smnzhu. 👍 makes sense

**Comment by smnzhu** [comment]
> Assuming that there is a consumer waiting to receive a notification and one thread sends a notification into the channel. Is receiving the notification an atomic operations such that any concurrent sending of notifications always observes the latest channel state?

@AlexHentschel I think you've shown that the above is not guaranteed. In other words, Golang's implementation of concurrent channel operations does not necessarily provide [linearizability](http://www.bailis.org/blog/linearizability-versus-serializability/#:~:text=linearizability%3A%20single-operation%2C%20single-object%2C%20real-time%20order). However, it does provide [serializability](http://www.bailis.org/blog/linearizability-versus-serializability/#:~:text=serializability%3A%20multi-operation%2C%20multi-object%2C%20arbitrary%20total%20order), ie that the execution of a set of channel operations is equivalent to *some* serial execution (total ordering) of the operations.

I don't think lack of linearizability in the example you provided (`TestNotifier_ManyConsumers`) is actually the issue though. Even if we assume that Golang's implementation is linearizable, we *still* can come up with race conditions which don't involve conflicting concurrent channel operations, as described [here](https://github.com/onflow/flow-go/pull/1045/files#r678068918). However, this is not an issue with the implementation of the `Notifier` imo, rather it's an issue with the test logic.

**Comment by AlexHentschel** [comment]
Thanks @smnzhu, I like your separation of aspects:
1. Do we desire [linearizability](http://www.bailis.org/blog/linearizability-versus-serializability/#:~:text=linearizability%3A%20single-operation%2C%20single-object%2C%20real-time%20order) (aka "atomic consistency")? Or is  [serializability](http://www.bailis.org/blog/linearizability-versus-serializability/#:~:text=serializability%3A%20multi-operation%2C%20multi-object%2C%20arbitrary%20total%20order) sufficient?

2. Do we have a problem in the tests (specifically `TestNotifier_ManyConsumers` and `TestNotifier_AllWorkProcessed`)
 
### Re 1: Do we desire linearizability or is serializability sufficient? 
@durkmurder has actually made quite a good argument that we desire atomic consistency, i.e. linearizability for system responsiveness:
* assume we have `P` producers and `X` consumers (workers)
  * producers are going to dup a _large number of work_ into the `pendingWorkQueue` 
  * there is a large number of consumers waiting to receive notifications and start processing the work
  * the producers triggering notifications and the consumers arriving at the `notifier` are concurrent operations
* With serializability, it means that _some_ order of events (other than physical event order according to the wall-clock) can occur. Specifically consider the following (probably frequently occurring) situation: 
   * Just a millisecond ago, the consumers were busy processing work. They then find that the `pendingWorkQueue` is empty. 
   * Then a new heap of work arrives. All consumers have already blocked on the notifier and are waiting to be woken up, when the _first_ producer sends a notification
   * With serializability, the system can behave _as if_ the producers dump _all_ their work into the queue and trigger notifications, and pretend that there is not worker waiting to be woken up. As a consequence, only one worker will receive a notification and must process _all_ the work by itself. 
*  In contrast, with linearizability, the software cannot temporally reorder events. The workers that are waiting for new work are guaranteed to be woken up, when new work arrives. 

Hence, I would argue that we desire linearizability and serializability is not sufficient for our use case. 

### Re 2: problem in the tests
Totally agree that it is hard to eliminate the race condition in the test. I like your pragmatic solution of just adding a `time.Sleep` (maybe even using 100ms to accommodate an overloaded CI instance)

**Comment by AlexHentschel** [comment]
Let me try to summarize this discussion and see if we have consensus 😅 

## Summary 

* For system responsiveness and load balancing, we **desire that the `notifier` guarantees atomic consistency, i.e. linearizability**.
* We conclude that **buffered channels provide serializability** for concurrent sends as well as concurrent receives. Hence, their guarantees are not strong enough. 

What do you think?

**Review by AlexHentschel — APPROVED** [review]
Thanks @smnzhu for your work on this and everyone for the _very insightful_ discussion. I learned a lot.

**Review by zhangchiqing — APPROVED** [review]
Great discussion. Thanks for the fix!

**Comment by smnzhu** [comment]
@AlexHentschel Thanks a lot for the detailed writeup of the example and for digging into the golang runtime implementation! I learned a lot from this discussion as well.
